### DISCUSSION TITLE
Livewire Protected Computed Property

### DISCUSSION THREAD

**Discussion by gofish543** [livewire/livewire#7823] [7. Bugs (archived)]
### Livewire version

 3.4.2

### Laravel version

v10.42.0

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

In prior versions of livewire (pre 3.4.2) you could have computed properties on protected methods. In the new version this was giving undefined exceptions.

This is intended? It was not documented anywhere.

### Repository or Wirebox URL

https://wirebox.app/b/qy82y

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by calebporzio** [comment]
Ok thank you for the report. I believe this recent PR caused the breakage: https://github.com/livewire/livewire/pull/7834

I reverted and will tag shortly. Please update Livewire to the latest version and see if that fixes it. Thanks

**Reply by luanfreitasdev** [reply]
Hello!. If a property was protected it should not be used as computed. Changing "protected" to "public" is easier than injecting a dependency. (The protected method is not documented). The Computed property worked as in version 2.

**Reply by devajmeireles** [reply]
I agree with you. It doesn't make sense for a protected method to be used as a computed property.

**Reply by calebporzio** [reply]
@luanfreitasdev i think both can be supported right?

**Reply by calebporzio** [reply]
Also I could be convinced, especially knowing it was that was in v2

**Reply by luanfreitasdev** [reply]
Hello @calebporzio. I think there are two ways out:

1 - Add the missing support:
* Allow a dynamic computed property to be called `protected` / `private`
* Allow dependency injection on the "`#[Computed]`" attribute (similar to dynamic).
* Document the possibility of calling `protected` / `private` methods in views.

2 – Revert the “Revert” and keep it the same as the dynamic method
* Document the inability to call `protected` / `private` methods in views.

**Note: We can now call a private method also in version 3.4.4**

**Reply by alexwass-lr** [reply]
Hey @calebporzio - sorry for the revival but bumped into this problem today and stumbled across this discussion and the relevant PR's. @luanfreitasdev has already covered my thoughts pretty well, but I'd like to add some extra info for your contemplation.

Anecdotally, I found myself automatically reaching for this feature as it's also [supported by actions](https://livewire.laravel.com/docs/actions#dependency-injection), so I was shocked when it errored. This might be a common assumption for newcomers too. I get an action could be thought of like a controller method which does have dependency injection, but normal methods don't and therefore injection should be done during `__construct`. Following this pattern, we could use `boot()` for the injection, it just doesn't feel right to me though - plus DI is just much neater.

Most importantly, computed properties are still properties, which have [always](https://laravel-livewire.com/docs/2.x/properties#introduction) [been documented](https://livewire.laravel.com/docs/properties) as needing to be public, I can't think of a reason why I'd need a non-public computed Livewire property? It's still passed to the view anyway.

At the moment the `ReflectionClass` in the `invade` helper allows us to evaluate non-public methods (_I've always found that feature of the Reflection API weird anyway - it's private for a reason_), but `Illuminate\Container\BoundMethod` tries to access the object from the container, thus enforcing expected scopes. I'm not sure there's a simple world here where we get all scope types, plus DI, unless we make another helper that is a some frankenstein of `$method->invoke` and `Livewire\ImplicitlyBoundMethod`.

**Reply by luanfreitasdev** [reply]
I fell into this same problem. That's why it's best to leave the method private for context reasons.
